### PR TITLE
annotate difference only

### PR DESCRIPTION
**Pull request by karamba228**

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -80,15 +81,62 @@ async function annotatePackage(packageName, filePath, lineNumber) {
...
 }
 
+// Fetch modified lines in the PR
+async function getModifiedLines(filePath) {
+    const token = core.getInput('github-token', { required: true });
+    const octokit = github.getOctokit(token);
+
+    const { context } = github;
+    const { owner, repo } = context.repo;
+    const pull_number = context.payload.pull_request.number;
```
Comment (srossross) [inline index.js:91]: Do we have access to `context.payload.pull_request.changed_files` ?

**Review by srossross — CHANGES_REQUESTED** [review]
please include a workflow that demos this capability

File: index.js (hunk 1 above)
Comment (srossross) [reply index.js:91]: or use context.payload.pull_request.base.ref to diff the files directly otherwise the user needs to set up a token and perms

Comment (karamba228) [reply index.js:91]: `context.payload.pull_request.number` just gets the general information about the PR, we get updated files and updated lines from the comand below: `octokit.rest.pulls.listFiles`

File: index.js [hunk 2]
```diff
@@ -80,15 +84,69 @@ async function annotatePackage(packageName, filePath, lineNumber) {
...
+// Fetch modified lines in the PR
+async function getModifiedLines(filePath) {
+    const { context } = github;
+    const baseRef = context.payload.pull_request.base.ref;
+    const headRef = context.payload.pull_request.head.ref;
+
+    try {
+        // Fetch and check out the base branch to diff against
+        await execPromise(`git fetch origin ${baseRef}`);
+        await execPromise(`git checkout ${baseRef}`);
```
Comment (srossross) [inline index.js:96]: Eon't this change the files to the base ref eg `origin/main`?

File: src/index.mjs [hunk 3]
```diff
@@ -190,13 +211,65 @@ async function processCondaEnvironment(filePath) {
     }
 }
 
+// Fetch modified lines in the PR
+async function getModifiedLinesFromCommitDiff(filePath) {
```
Comment (srossross) [inline src/index.mjs:215]: Can you move this to its own file `diffUtils.js`?

File: src/index.mjs [hunk 4]
```diff
@@ -190,13 +211,65 @@ async function processCondaEnvironment(filePath) {
...
+    } catch (error) {
+        core.setFailed(`Error getting modified lines from commit diff: ${error.message}`);
+        return [];
+    }
+}
+
 async function run() {
     const ecosystem = core.getInput('package-ecosystem', { required: true });
+    const annotateModifiedOnly = core.getInput('annotate-modified-only') === 'true';
+    let modifiedLines = [];
```
Comment (srossross) [inline src/index.mjs:252]: if you set this to an array of the file line count you can simplify below

```
        if (modifiedLines.length > 0) {
            await processLines('requirements.txt', modifiedLines, 'pip');
        } else {
            await processPipRequirements('requirements.txt');
        }
```

to just:
```
        await processLines('requirements.txt', modifiedLines, 'pip');
        
```

File: src/index.mjs [hunk 5]
```diff
@@ -190,13 +211,65 @@ async function processCondaEnvironment(filePath) {
     }
 }
 
+// Fetch modified lines in the PR
+async function getModifiedLinesFromCommitDiff(filePath) {
+    try {
+        const { stdout, stderr } = await execPromise(`git diff HEAD^ HEAD -- ${filePath}`);
```
Comment (srossross) [inline src/index.mjs:217]: if you add the `--unified=0` flag this will not output any context lines

File: .github/workflows/test-cli.yml [hunk 6]
```diff
@@ -7,6 +7,8 @@ jobs:
     runs-on: ubuntu-latest
     steps:
       - uses: actions/checkout@v3
+        with:
+          fetch-depth: 2
```
Comment (srossross) [inline .github/workflows/test-cli.yml:11]: Won't all users have to do this? how will this fail if it is forgotten? I'd prefer to do this in a robust way that does not require external config.

Also does this work for pull requests?

File: src/index.mjs (hunk 5 above)
Comment (srossross) [reply src/index.mjs:217]: also why are you doing `diff HEAD^ HEAD`? This is supposed to be for pull request won't that only diff the last commit?

**Review by srossross — CHANGES_REQUESTED** [review]

File: src/index.mjs (hunk 5 above)
Comment (srossross) [reply src/index.mjs:217]: you might need to diff with `{context.payload.pull_request.base.ref}`

File: src/index.mjs [hunk 7]
```diff
@@ -190,13 +212,71 @@ async function processCondaEnvironment(filePath) {
...
     } else if (ecosystem === 'conda') {
-        await processCondaEnvironment('environment.yml');
+        if (annotateModifiedOnly) {
+            modifiedLines = await getModifiedLines('environment.yml');
+        }
+        if (modifiedLines.length > 0) {
+            await processLines('environment.yml', modifiedLines, 'conda');
+        } else {
+            await processCondaEnvironment('environment.yml');
+        }
```
Comment (srossross) [inline src/index.mjs:279]: can you remove these if statements as per https://github.com/openteamsinc/actions/pull/8/files#r1775773852

File: src/index.mjs [hunk 8]
```diff
@@ -190,13 +212,71 @@ async function processCondaEnvironment(filePath) {
     }
 }
 
+// Fetch modified lines in the PR
+async function getModifiedLines(filePath) {
```
Comment (srossross) [inline src/index.mjs:216]: can you move this to its own file as per https://github.com/openteamsinc/actions/pull/8#discussion_r1775771704

File: src/index.mjs [hunk 9]
```diff
@@ -190,13 +212,71 @@ async function processCondaEnvironment(filePath) {
...
+    const { context } = github;
+    const baseRef = context.payload.pull_request.base.ref;
+
+    try {
+        // Fetch the base branch to ensure we have the latest state of baseRef locally
+        await execPromise(`git fetch origin ${baseRef}`);
```
Comment (srossross) [inline src/index.mjs:222]: please check that baseRef is not null and provide the user with a good error message

**Review by srossross — CHANGES_REQUESTED** [review]
there were some un-addressed comments from the previous review and the baseRef might not exist

**Review by srossross — APPROVED** [review]

**Pull request merged**